text/0000-supertrait-item-shadowing-v2.md

```diff
@@ -0,0 +1,77 @@
+- Feature Name: `supertrait_item_shadowing`
+- Start Date: 2024-05-04
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+When name resolution encounters an ambiguity between 2 trait methods, if one trait is a sub-trait of the other then select that method instead of reporting an ambiguity error.
+
+# Motivation
+[motivation]: #motivation
+
+
+The libs-api team would like to stabilize `Iterator::intersperse` but has a problem. The `itertools` crate already has:
+
+```rust
+// itertools
+trait Itertools: Iterator {
+    fn intersperse(self, element: Self::Item) -> Intersperse<Self>;
+}
+```
+
+This method is used in crates with code similar to the following:
+
+```rust
+use core::iter::Iterator; // Implicit import from prelude
+
+use itertools::Itertools as _;
+
+fn foo() -> impl Iterator<Item = &'static str> {
+    "1,2,3".split(",").intersperse("|")
+    // ^ This is ambiguious: it could refer to Iterator::intersperse or Itertools::intersperse
+}
+```
+
+This code actually works today because `intersperse` is an unstable API, which works because the compiler already has [logic](https://github.com/rust-lang/rust/pull/48552) to prefer stable methods over unstable methods when an amiguity occurs.
+
+Attempts to stabilize `intersperse` have failed with a large number of regressions [reported by crater](https://github.com/rust-lang/rust/issues/88967) which affect many popular crates. Even if these were to be manually corrected (since ambiguity is considered allowed breakage) we would have to go through this whole process again every time a method from `itertools` is uplifted to the standard library.
+
+# Proposed solution
+[proposed-solution]: #proposed-solution
+
+This RFC proposes to change name resolution to resolve the ambiguity in the following specific circumstances:
+- All method candidates are trait methods. (Inherent methods are already prioritized over trait methods)
+- One trait is transitively a sub-trait of all other traits in the candidate list.
+
+When this happens, the sub-trait method is selected instead of reporting an ambiguity error.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This behavior can be surprising: adding a method to a sub-trait can change which function is called in unrelated code. A lint could be emitted to warn users about the potential ambiguity.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+If we choose not to accept this RFC then there doesn't seem to be a reasonable path for adding new methods to the `Iterator` trait if such methods are already provided by `itertools` without a lot of ecosystem churn.
```

Comment:
Why is picking a different name not a reasonable path? The unavoidable bikeshedding around a new name is annoying, but it seems to me like a small, one-time cost compared to the permanent additional language complexity of this feature.

I also wonder how many times we anticipate to run into this problem in the future. Are there more examples aside from `Itertools::intersperse`? If we only ran into this problem once within 9 years of Rust being stable, the benefit of this feature seems very limited.

Reply:
Picking a new name also has disadvantages other than than the time taken to pick it, users then have to learn the new method name and that it is semantically identical to the `Itertools::intersperse()` yet has a different name. This is not a one-time cost as it will effect all future users when, for example, searching docs for this method.

Reply:
> users then have to learn the new method name

I don't see the problem. People who are using `intersperse` from `itertools` shouldn't expect the standard library to use the exact same name. And I assume few people rely on the unstable version in `std` for production code, it would be unwise to opt into breaking changes for what amounts to a small ergonomics improvement (which a library can also provide).

For regular Rust users, the function doesn't exist right now. In the future it may - its name is an open question.

> This would be the case for all methods stabilized from `itertools` to `std` which might be quite a few

I agree that this is something to consider, we don't want to have to pick the "second best" name for many functions in `std`. But first we should think concretely about which methods from `itertools` might actually make the jump into `std`. itertools has been around for a long time and I'm not aware of a big push to upstream many of its methods. Which indicates to me, there isn't that much need. But I'm happy to be convinced otherwise. Examples from other libraries besides itertools count as well.

Reply:
The issue is that because itertools is used widely, std *cannot* upstream methods from itertools without causing lots of breakage in all crates currently using itertools. This is a perverse incentive which forces the "second best" issue you mentioned, and it's not limited to itertools, it happens whenever std lacks a function, someone implements an extension trait to add it in a crate (as they should), and everyone picks it up because it is really useful (as they should).

Reply:
I understand that. The reason I'm pushing back is that to me, the feature doesn't seem to align with Rust's design principles. This new implicit default doesn't seem obviously correct to me. If I call a method that has two implementations, I would generally prefer the compiler yell at me rather than pick one without telling me. That's why I think we should be certain that we'll make good use of this feature before adding it.

But I'll admit this is a theoretical objection. In practice, the problem may never show up and then it's fine to add the feature. Pragmatism comes first. I guess I just agree with [this comment](https://github.com/rust-lang/rust/issues/89151#issuecomment-2075272805). We should think about edge cases where this could go wrong.

Reply:
> I would generally prefer the compiler yell at me rather than pick one without telling me

The rfc mentions this:

> This behavior can be surprising: adding a method to a sub-trait can change which function is called in unrelated code. A lint could be emitted to warn users about the potential ambiguity.

Reply:
Yeah, a lint would make sense if the migration-situation with intersperse is the only use case we expect this feature to be used in practice. But maybe some users will use the shadowing to implement a form of specialization? Or any other unrelated use case I can't think of right now. If that happens, it might lead to a discussion about whether the lint should be enabled by default or not. And if it ends up allow-by-default, it loses much of its value.